pkg/networkpolicy/packet.go

```diff
@@ -29,59 +33,111 @@ func (p packet) String() string {
+		// will never be passed via the nfqueue. Packets are
+		// re-assembled by the kernel. Please see:
+		// https://unix.stackexchange.com/questions/650790/unwanted-defragmentation-of-forwarded-ipv4-packets
 	case 6:
 		t.family = v1.IPv6Protocol
-		hdrlen = 40
-		if len(b) < hdrlen+4 {
-			return t, fmt.Errorf("payload to short, received %d expected at least %d", len(b), hdrlen+4)
+		if len(b) < 48 {
+			// 40 is the minimum length of an IPv6 header, and 8 is
+			// the minimum lenght of an extension or L4 header
+			return t, ErrorTooShort
 		}
 		t.srcIP = make(net.IP, net.IPv6len)
 		copy(t.srcIP, b[8:24])
 		t.dstIP = make(net.IP, net.IPv6len)
 		copy(t.dstIP, b[24:40])
-		// NextHeader (not extension headers supported)
-		protocol = int(b[6])
+		// Handle extension headers.
+		nxtHeader = int(b[6])
+		l4offset = 40
+		for nxtHeader == syscall.IPPROTO_DSTOPTS || nxtHeader == syscall.IPPROTO_HOPOPTS || nxtHeader == syscall.IPPROTO_ROUTING {
+			// These headers have a lenght in 8-octet units, not
+			// including the first 8 octets
+			nxtHeader = int(b[l4offset])
+			l4offset += (8 + int(b[l4offset+1])*8)
+			// Now l4offset points to either another extension header,
+			// or an L4 header. So we must have at least 8 byte data
+			// after this (minimum extension header size)
+			if (l4offset + 8) >= len(b) {
+				return t, ErrorTooShort
+			}
+		}
+		if nxtHeader == syscall.IPPROTO_FRAGMENT {
+			// Only the first fragment has the L4 header
+			fragOffset := int(binary.BigEndian.Uint16(b[l4offset+2 : l4offset+4]))
+			if fragOffset&0xfff8 == 0 {
+				nxtHeader = int(b[l4offset])
+				l4offset += 8
+				// Here it's assumed that the fragment is the last
+				// extension header before the L4 header. But more
+				// IPPROTO_DSTOPTS are allowed by the recommended order.
+				// TODO: handle extra IPPROTO_DSTOPTS.
+			} else {
+				// If this is NOT the first fragment, we have no L4
+				// header and the payload begins after this
+				// header. Return a packet with t.proto unset
+				return t, nil
+			}
+		}
+		protocol = nxtHeader
 	default:
-		return t, fmt.Errorf("unknown versions %d", version)
+		return t, fmt.Errorf("unknown version %d", version)
 	}
 
-	var dataOffset int
 	switch protocol {
-	case 6:
+	case syscall.IPPROTO_TCP:
 		t.proto = v1.ProtocolTCP
-		dataOffset = int(b[hdrlen+12] >> 4) // data offset
-	case 17:
+		dataOffset := int(b[l4offset+12]>>4) * 4
+		if dataOffset < 20 {
+			return t, ErrorCorrupted
+		}
+		payloadOffset = l4offset + dataOffset
+	case syscall.IPPROTO_UDP:
 		t.proto = v1.ProtocolUDP
-		dataOffset = hdrlen + 8 // data starts after
-	case 132:
+		payloadOffset = l4offset + 8
+	case syscall.IPPROTO_SCTP:
 		t.proto = v1.ProtocolSCTP
-		dataOffset = hdrlen + 8
+		payloadOffset = l4offset + 8
 	default:
+		// This "should not happen" since only { tcp, udp, sctp } are
+		// routed to the nfqueue
 		return t, fmt.Errorf("unknown protocol %d", protocol)
```

Comment:
we are going to route any protocol here, so we can emit verdict on icmp per example, more on this in https://github.com/kubernetes-sigs/kube-network-policies/pull/51/files#r1681040192